// @ebazdev/order/build/models/order-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderTemplate = exports.TemplateType = void 0;
const mongoose_1 = require("mongoose");
const mongoose_update_if_current_1 = require("mongoose-update-if-current");
var TemplateType;
(function (TemplateType) {
    TemplateType["Supplier"] = "supplier";
    TemplateType["Merchant"] = "merchant";
})(TemplateType || (exports.TemplateType = TemplateType = {}));
const orderTemplateProductSchema = new mongoose_1.Schema({
    id: {
        type: mongoose_1.Schema.Types.ObjectId,
        required: true,
        ref: "Product",
    },
    quantity: {
        type: Number,
        required: true,
    },
}, { _id: false });
const orderTemplateSchema = new mongoose_1.Schema({
    type: { type: String, enum: Object.values(TemplateType), required: true },
    supplierId: {
        type: mongoose_1.Schema.Types.ObjectId,
        required: true,
        ref: "Customer",
    },
    merchantId: {
        type: mongoose_1.Schema.Types.ObjectId,
        required: false,
        ref: "Customer",
    },
    products: [orderTemplateProductSchema],
    name: {
        type: String,
        required: true,
    },
    image: {
        type: String,
        required: false,
    },
    color: {
        type: String,
        required: false,
    },
}, {
    timestamps: true,
    toJSON: {
        transform(doc, ret) {
            ret.id = ret._id;
            delete ret._id;
        },
    },
});
orderTemplateSchema.set("versionKey", "version");
orderTemplateSchema.plugin(mongoose_update_if_current_1.updateIfCurrentPlugin);
const OrderTemplate = (0, mongoose_1.model)("OrderTemplate", orderTemplateSchema);
exports.OrderTemplate = OrderTemplate;
